consensus/chain_selector.go

```diff
@@ -17,23 +19,96 @@ import (
 	"github.com/filecoin-project/go-filecoin/types"
 )
 
+// NewECV is the constant V defined in the EC spec.
+const NewECV uint64 = 2
+
+// PI is the multiplicand in the null penalty term
+const PI float64 = 0.87
+
+// NullThresh is the min number of null rounds before the penalty kicks in
+const NullThresh = 3
+
+// ECV is the constant V defined in the EC spec.
+const ECV uint64 = 10
+
+// ECPrM is the power ratio magnitude defined in the EC spec.
+const ECPrM uint64 = 100
+
+// ChainSelector weighs and compares chains according to the Storage Power
+// Consensus Protocol
 type ChainSelector struct {
-	cstore *hamt.CborIpldStore
+	cstore     *hamt.CborIpldStore
 	actorState SnapshotGenerator
 	genesisCid cid.Cid
 }
 
 // NewChainSelector is the constructor for chain selection module.
 func NewChainSelector(cs *hamt.CborIpldStore, actorState SnapshotGenerator, gCid cid.Cid) *ChainSelector {
 	return &ChainSelector{
-		cstore:            cs,
-		actorState:        actorState,
-		genesisCid:        gCid,
+		cstore:     cs,
+		actorState: actorState,
+		genesisCid: gCid,
+	}
+}
+
+// NewWeight returns the EC weight of this TipSet in uint64 encoded fixed point
+// representation.
+//
+// w(i) = w(i-1) + (PI)^(P_n) * [V * num_blks + X ]
+// P_n(n) = if n < 3:0 else: n, n is number of null rounds
+// X = log_2(total_storage(pSt))
+func (c *ChainSelector) NewWeight(ctx context.Context, ts types.TipSet, pSt state.Tree) (uint64, error) {
+	ctx = log.Start(ctx, "Expected.Weight")
+	log.LogKV(ctx, "Weight", ts.String())
+	if ts.Len() == 1 && ts.At(0).Cid().Equals(c.genesisCid) {
+		return uint64(0), nil
+	}
+	// Compute parent weight.
+	parentW, err := ts.ParentWeight()
+	if err != nil {
+		return uint64(0), err
 	}
+
+	w, err := types.FixedToBig(parentW)
+	if err != nil {
+		return uint64(0), err
+	}
+
+	// Each block adds ECV to the weight's inner term
+	innerTerm := new(big.Float)
+	floatECV := new(big.Float).SetInt64(int64(NewECV))
+	floatNumBlocks := new(big.Float).SetInt64(int64(ts.Len()))
+	innerTerm.Mul(floatECV, floatNumBlocks)
+
+	// Add X to the weight's inner term
+	powerTableView := c.createPowerTableView(pSt)
+	totalBytes, err := powerTableView.Total(ctx)
+	if err != nil {
+		return uint64(0), err
+	}
+	roughLogTotalBytes := new(big.Float).SetInt64(int64(totalBytes.BigInt().BitLen()))
+	innerTerm.Add(innerTerm, roughLogTotalBytes)
+
+	// Attenuate weight by the number of tickets
+	numTickets := len(ts.At(0).Tickets)
+	P := new(big.Float).SetInt64(int64(1))
+	if numTickets >= NullThresh {
+		bigPI := new(big.Float).SetFloat64(PI)
+		// P = PI^numNull
+		for i := 0; i < numTickets; i++ {
```

Comment:
It is intended to have at least PI to the power of NullThresh (0.87^3 = 0.6585) as the ratio to attenuate weight? This happens when there are 2 null blocks between tipsets. It looks too much, while there is no attenuation when one null block exists between tipsets. 

I would suggest changing it to: 
```golang
          for i := 0; i < numTickets-1; i++ {
              ... 
         }
```

Reply:
Thanks for looking at this closely @steven004.  From the version of the spec I implemented against 
- `Weight = ParentWeight + P(V*NumBlocksMinedInThisRound + X)`
- `P will be equal to some term P_i^P_n`
- `P_n = the number of tickets in this block's Ticket array if that number is  >= 3, 0 otherwise.`

So this implementation is up to that version of the spec.

Unfortunately I can't link this to you because rapid changes to the spec have shifted the weighting function definition yet again :(  

However this might mean that your concerns are addressed by [the new function](https://filecoin-project.github.io/specs/#chain-selection).  Please leave github issues in [the specs repo](https://github.com/filecoin-project/specs) for concerns about that latest speced weighting function.